File: scripts/stitch_esri.py
from osgeo import gdal
import os
import glob
import mercantile
import rasterio
from rasterio.transform import from_bounds
from rasterio.crs import CRS
import warnings

gdal.UseExceptions()
warnings.filterwarnings("ignore", category=rasterio.errors.NotGeoreferencedWarning)

TILE_DIR = "data/esri_tiles"
OUTPUT_COG = "data/esri_test_roi_cog.tif"


def georeference_tile_with_rasterio(tile_path, x, y, zoom):
    tile = mercantile.Tile(x, y, zoom)
    bounds = mercantile.bounds(tile)  # Returns LngLatBbox(west, south, east, north)
    georef_path = tile_path.replace(".png", "_geo.tif")
    with rasterio.open(tile_path) as src:
        transform = from_bounds(
            bounds.west, bounds.south, bounds.east, bounds.north, src.width, src.height
        )
        profile = src.profile.copy()
        profile.update(
            {"driver": "GTiff", "crs": CRS.from_epsg(4326), "transform": transform}
        )
        with rasterio.open(georef_path, "w", **profile) as dst:
            for i in range(1, src.count + 1):
                dst.write(src.read(i), i)
    return georef_path


def create_cog_from_tiles():
    tiles = glob.glob(f"{TILE_DIR}/tile_*.png")
    if not tiles:
        print(f"No tiles found in {TILE_DIR}")
        return False
    print(f"Found {len(tiles)} tiles")
    georef_tiles = []
    for tile_path in tiles:
        basename = os.path.basename(tile_path)
        parts = basename.replace(".png", "").split("_")
        if len(parts) != 4:
            continue
        zoom = int(parts[1])
        x = int(parts[2])
        y = int(parts[3])
        print(f"Georeferencing tile {x}/{y}/{zoom}...")
        try:
            georef_path = georeference_tile_with_rasterio(tile_path, x, y, zoom)
            georef_tiles.append(georef_path)
        except Exception as e:
            print(f"Failed to georeference {tile_path}: {e}")
    if not georef_tiles:
        print("No tiles could be georeferenced")
        return False
    vrt_file = "data/esri_test_roi.vrt"
    try:
        print("Creating VRT...")
        gdal.BuildVRT(vrt_file, georef_tiles)
        print("Creating COG...")
        gdal.Translate(
            OUTPUT_COG, vrt_file, format="COG", creationOptions=["COMPRESS=DEFLATE"]
        )
        print(f"Created COG: {OUTPUT_COG}")
        return True
    finally:
        if os.path.exists(vrt_file):
            os.remove(vrt_file)
        for georef_tile in georef_tiles:
            if os.path.exists(georef_tile):
                os.remove(georef_tile)


if __name__ == "__main__":
    create_cog_from_tiles()

File: scripts/stitch_esri.js
import { execSync } from "child_process";
import { readdirSync, existsSync, unlinkSync } from "fs";
import { join } from "path";
import yaml from "js-yaml";
import SphericalMercator from "sphericalmercator";

async function main() {
  const configPath = join(__dirname, "..", "config.yaml");
  const configFile = await Bun.file(configPath).text();
  const config = yaml.load(configFile);
  const aoiName = config.aoi.current;
  const aoiConfig = config.aoi[aoiName];
  if (!aoiConfig || aoiConfig.source !== "esri") {
    throw new Error(`No ESRI config for AOI: ${aoiName}`);
  }
  const TILE_DIR = join(aoiConfig.output_dir, "inputs", "esri");
  const OUTPUT_COG = join(
    aoiConfig.output_dir,
    "intermediates",
    "stitched-esri.tif"
  );
  const zoom = aoiConfig.scale;
  if (!existsSync(TILE_DIR)) {
    console.error(
      `Tile directory ${TILE_DIR} not found. Run download script for '${aoiName}' first.`
    );
    process.exit(1);
  }
  const merc = new SphericalMercator({ size: 256 });

  function getTileInfo(filename) {
    const parts = filename.replace(".jpg", "").split("_"); // .jpg for ESRI
    if (parts.length !== 4 || parseInt(parts[1]) !== zoom) return null; // Filter for config zoom
    return {
      zoom: parseInt(parts[1]),
      x: parseInt(parts[2]),
      y: parseInt(parts[3]),
      path: join(TILE_DIR, filename),
    };
  }

  function getTileBounds(x, y, zoom) {
    const bounds = merc.bbox(x, y, zoom);
    return {
      west: bounds[0],
      south: bounds[1],
      east: bounds[2],
      north: bounds[3],
    };
  }

  function georeferenceTile(tile) {
    const bounds = getTileBounds(tile.x, tile.y, tile.zoom);
    const tempPath = tile.path.replace(".jpg", "_temp.tif");
    const georefPath = tile.path.replace(".jpg", "_geo.tif");
    console.log(`Georeferencing tile ${tile.x}/${tile.y}/${tile.zoom}...`);
    try {
      const cmd1 = `gdal_translate -of GTiff -gcp 0 0 ${bounds.west} ${bounds.north} -gcp 256 0 ${bounds.east} ${bounds.north} -gcp 0 256 ${bounds.west} ${bounds.south} -gcp 256 256 ${bounds.east} ${bounds.south} -a_srs EPSG:4326 "${tile.path}" "${tempPath}"`;
      execSync(cmd1, { stdio: "pipe" });
      const cmd2 = `gdalwarp -of GTiff -t_srs EPSG:4326 "${tempPath}" "${georefPath}"`;
      execSync(cmd2, { stdio: "pipe" });
      return georefPath;
    } finally {
      try {
        if (existsSync(tempPath)) {
          unlinkSync(tempPath);
        }
      } catch (e) {
        console.warn(`Could not delete temp file ${tempPath}: ${e.message}`);
      }
    }
  }

  const files = readdirSync(TILE_DIR).filter((f) => f.endsWith(".jpg")); // .jpg for ESRI
  console.log(`Found ${files.length} tiles for AOI '${aoiName}'`);
  const tiles = files.map(getTileInfo).filter(Boolean);
  const georefTiles = [];
  for (const tile of tiles) {
    try {
      const georefPath = georeferenceTile(tile);
      georefTiles.push(georefPath);
    } catch (error) {
      console.error(`Failed to georeference ${tile.path}: ${error.message}`);
    }
  }
  if (georefTiles.length === 0) {
    console.error("No tiles could be georeferenced");
    process.exit(1);
  }
  const vrtFile = join(aoiConfig.output_dir, `esri_${aoiName}_roi.vrt`);
  try {
    console.log("Creating VRT and COG...");
    execSync(
      `gdalbuildvrt "${vrtFile}" ${georefTiles.map((p) => `"${p}"`).join(" ")}`
    );
    execSync(
      `gdal_translate -of COG -co COMPRESS=DEFLATE "${vrtFile}" "${OUTPUT_COG}"`
    );
    console.log(`Created COG: ${OUTPUT_COG}`);
  } finally {
    console.log("Cleaning up intermediate files...");
    try {
      if (existsSync(vrtFile)) {
        unlinkSync(vrtFile);
      }
    } catch (e) {
      console.warn(`Could not delete VRT ${vrtFile}: ${e.message}`);
    }
    for (const georefTile of georefTiles) {
      try {
        if (existsSync(georefTile)) {
          unlinkSync(georefTile);
        }
      } catch (e) {
        console.warn(`Could not delete ${georefTile}: ${e.message}`);
      }
    }
    console.log("Cleanup complete");
  }
}

console.log(`Stitching tiles for AOI: ${aoiName}`);
main().catch(console.error);
